packages/cosmic-proto/README.md

```diff
@@ -1,9 +1,280 @@
-# Protobuf Stubs for Cosmic Swingset
+# @agoric/cosmic-proto
 
-This package exports JavaScript and TypeScript generated from Cosmic Swingset
-Protobuf IDL so that clients can communicate with the Agoric chain.
+<p align="center">
+  <img src="https://user-images.githubusercontent.com/545047/188804067-28e67e5e-0214-4449-ab04-2e0c564a6885.svg" width="80"><br />
+    Protobuf interfaces for Agoric on Cosmos
+</p>
 
-Package exports are defined in `exports` and also made available in the
-`swingset/` file path for backwards compatibility.
 
-See also [MAINTAINERS.md](MAINTAINERS.md).
+## install
+
+```sh
+npm install @agoric/cosmic-proto
+```
+## Table of contents
+
+- [Usage](#usage)
+    - [RPC Clients](#rpc-clients)
+    - [Composing Messages](#composing-messages)
+        - Cosmos, CosmWasm, and IBC
+            - [CosmWasm](#cosmwasm-messages)
+            - [IBC](#ibc-messages)
+            - [Cosmos](#cosmos-messages)
+- [Wallets and Signers](#connecting-with-wallets-and-signing-messages)
+    - [Stargate Client](#initializing-the-stargate-client)
+    - [Creating Signers](#creating-signers)
+    - [Broadcasting Messages](#broadcasting-messages)
+- [Advanced Usage](#advanced-usage)
+- [Developing](#developing)
+- [Credits](#credits)
+
+## Usage
+### RPC Clients
+
+```js
+import { agoric } from '@agoric/cosmic-proto';
+
+const { createRPCQueryClient } = agoric.ClientFactory; 
+const client = await createRPCQueryClient({ rpcEndpoint: RPC_ENDPOINT });
+
+// now you can query the cosmos modules
+const balance = await client.cosmos.bank.v1beta1
+    .allBalances({ address: 'agoric1addresshere' });
+
+// you can also query the agoric modules
+const swingsetParams = await client.agoric.swingset.params()
+```
+
+### Composing Messages
+
+Import the `agoric` object from `@agoric/cosmic-proto`. 
+
+```js
+import { agoric } from '@agoric/cosmic-proto';
+
+const {
+    installBundle,
+} = agoric.exchange.v1beta1.MessageComposer.withTypeUrl;
```

Comment:
```suggestion
} = agoric.swingset.v1beta1.MessageComposer.withTypeUrl;
```

Oddly, typescript complains if I don't include a value for `bundle`:

<details>

<summary>

```
Argument of type '{ compressedBundle: Uint8Array; submitter: Uint8Array; uncompressedSize: bigint; }' is not assignable to parameter of type 'MsgInstallBundle'.
Property 'bundle' is missing in type '{ compressedBundle: Uint8Array; submitter: Uint8Array; uncompressedSize: bigint; }' but required in type 'MsgInstallBundle'.
```
</summary>

```ts
import { agoric } from '@agoric/cosmic-proto';
import { fromBech32 } from "@cosmjs/encoding";

function foo() {
  const {
    installBundle,
  } = agoric.swingset.MessageComposer.withTypeUrl; // same with `.fromPartial`
  const compressedBundle = new Uint8Array(0);
  const submitter = fromBech32('agoric1').data;
  const uncompressedSize = 1000n;
  const msg = installBundle({ compressedBundle, submitter, uncompressedSize });
}

foo();
```
</details>

L125 of `proto/agoric/swingset/msgs.proto` states: `Either bundle or compressed_bundle will be set.`

The exported interface from current @agoric/cosmic-proto seems to be the same, and `.encode` demands both `bundle` and `compressed_bundle` so this is not a regression.

Reply:
I'm not versed enough on .proto to know if it can be fixed there, a quick google search tells  me conditional logic / mutual exclusivity cannot be captured in the message definition, only at the application level. 

Worth a manual patch for the type?

```ts
export type MsgInstallBundle = {
  submitter: Uint8Array;
  uncompressedSize: bigint;
} & (
  | { bundle: string; compressedBundle?: never }
  | { bundle?: never; compressedBundle: Uint8Array }
);
```

Reply:
I thought maybe flipping the `useUnionTypes` option would do it but, but no change in output.

> Worth a manual patch for the type?

We want this to work the same as Telescope elsewhere. If we need it, let's find a way to do it in Telescope, even if that means a PR upstream.

Reply:
> a quick google search tells me conditional logic / mutual exclusivity cannot be captured in the message definition

chatgpt is now telling me that `oneof` is a thing. Maybe we should change our .proto definition?

```diff
diff --git a/golang/cosmos/proto/agoric/swingset/msgs.proto b/golang/cosmos/proto/agoric/swingset/msgs.proto
index fcd96c1ea..6d9090922 100644
--- a/golang/cosmos/proto/agoric/swingset/msgs.proto
+++ b/golang/cosmos/proto/agoric/swingset/msgs.proto
@@ -113,26 +113,31 @@ message MsgProvisionResponse {}

 // MsgInstallBundle carries a signed bundle to SwingSet.
 message MsgInstallBundle {
-    string bundle = 1 [
-        (gogoproto.jsontag)    = "bundle",
-        (gogoproto.moretags)   = "yaml:\"bundle\""
-    ];
+    // oneof is used to specify that either bundle or compressed_bundle can be set
+    oneof bundle_option {
+        string bundle = 1 [
+            (gogoproto.jsontag)    = "bundle",
+            (gogoproto.moretags)   = "yaml:\"bundle\""
+        ];
+        bytes compressed_bundle = 3 [
+            (gogoproto.jsontag)    = "compressedBundle",
+            (gogoproto.moretags)   = "yaml:\"compressedBundle\""
+        ];
+    }
+
     bytes submitter = 2 [
         (gogoproto.casttype)   = "github.com/cosmos/cosmos-sdk/types.AccAddress",
         (gogoproto.jsontag)    = "submitter",
         (gogoproto.moretags)   = "yaml:\"submitter\""
     ];
-    // Either bundle or compressed_bundle will be set.
-    // Default compression algorithm is gzip.
-    bytes compressed_bundle = 3 [
-        (gogoproto.jsontag)    = "compressedBundle",
-        (gogoproto.moretags)   = "yaml:\"compressedBundle\""
-    ];
+
     // Size in bytes of uncompression of compressed_bundle.
     int64 uncompressed_size = 4 [
-        (gogoproto.jsontag) = "uncompressedSize"
+        (gogoproto.jsontag) = "uncompressedSize",
+        (gogoproto.moretags)   = "yaml:\"uncompressedSize\""
     ];
 }
+}
```

Reply:
I'd be happy with this proto change if it's guaranteed not to be a breaking one.  Our golang CI would verify that.

Reply:
Welp, seems like a breaking change: https://github.com/Agoric/agoric-sdk/actions/runs/8113540583/job/22177076626?pr=9017#step:4:69